File: packages/react-native-nitro-image/android/src/main/java/com/margelo/nitro/image/HybridImageFactory.kt
package com.margelo.nitro.image

import android.annotation.SuppressLint
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import androidx.annotation.Keep
import com.facebook.common.internal.DoNotStrip
import com.facebook.react.bridge.ReactApplicationContext
import com.madebyevan.thumbhash.ThumbHash
import com.margelo.nitro.NitroModules
import com.margelo.nitro.core.ArrayBuffer
import com.margelo.nitro.core.Promise
import java.nio.ByteBuffer

@DoNotStrip
@Keep
class HybridImageFactory: HybridImageFactorySpec() {
    @SuppressLint("DiscouragedApi")
    override fun loadFromResources(name: String): HybridImageSpec {
        val context = NitroModules.applicationContext ?: throw Error("No context!")
        // Look up ID via it's name
        val rawResourceId: Int = context.resources
            .getIdentifier(name, "drawable", context.packageName)
        if (rawResourceId == 0) {
            // It's bundled into the Android resources/assets
            context.assets.open(name).use { stream ->
                val bitmap = BitmapFactory.decodeStream(stream)
                return HybridImage(bitmap)
            }
        } else {
            // For assets bundled with 'require' instead of linked, they are bundled into `res/raw` in release mode
            context.resources.openRawResource(rawResourceId).use { stream ->
                val bitmap = BitmapFactory.decodeStream(stream)
                return HybridImage(bitmap)
            }
        }
    }
    override fun loadFromResourcesAsync(name: String): Promise<HybridImageSpec> {
        return Promise.async { loadFromResources(name) }
    }

    override fun loadFromSymbol(symbolName: String): HybridImageSpec {
        throw Error("ImageFactory.loadFromSymbol(symbolName:) is not supported on Android!")
    }

    override fun loadFromRawPixelData(data: RawPixelData): HybridImageSpec {
        val bitmap = bitmapFromRawPixelData(data)
        return HybridImage(bitmap)
    }
    override fun loadFromRawPixelDataAsync(data: RawPixelData): Promise<HybridImageSpec> {
        val bufferCopy = data.buffer.copyIfNotOwner()
        val dataCopy = RawPixelData(bufferCopy, data.width, data.height, data.pixelFormat)
        return Promise.async { loadFromRawPixelData(dataCopy) }
    }

    private fun loadFromEncodedBytes(bytes: ByteArray): HybridImageSpec {
        val bitmap = BitmapFactory.decodeByteArray(bytes, 0, bytes.size)
        return HybridImage(bitmap)
    }
    override fun loadFromEncodedImageData(data: EncodedImageData): HybridImageSpec {
        val bytes = data.buffer.toByteArray()
        return loadFromEncodedBytes(bytes)
    }
    override fun loadFromEncodedImageDataAsync(data: EncodedImageData): Promise<HybridImageSpec> {
        val bytes = data.buffer.toByteArray()
        return Promise.async { loadFromEncodedBytes(bytes) }
    }

    override fun loadFromFile(filePath: String): HybridImageSpec {
        val bitmap = BitmapFactory.decodeFile(filePath)
        return HybridImage(bitmap)
    }

    override fun loadFromFileAsync(filePath: String): Promise<HybridImageSpec> {
        return Promise.async { loadFromFile(filePath) }
    }

    private fun loadFromThumbHash(thumbHashBytes: ByteArray): HybridImage {
        val rgba = ThumbHash.thumbHashToRGBA(thumbHashBytes)

        val bitmap = Bitmap.createBitmap(rgba.width, rgba.height, Bitmap.Config.ARGB_8888)
        val buffer = ByteBuffer.wrap(rgba.rgba)
        bitmap.copyPixelsFromBuffer(buffer)
        return HybridImage(bitmap)
    }

    override fun loadFromThumbHash(thumbhash: ArrayBuffer): HybridImageSpec {
        // copyIfNeeded can be false since we are running this synchronously
        val bytes = thumbhash.toByteArray()
        return loadFromThumbHash(bytes)
    }

    override fun loadFromThumbHashAsync(thumbhash: ArrayBuffer): Promise<HybridImageSpec> {
        // copyIfNeeded needs to be true since we switch threads
        val bytes = thumbhash.toByteArray()
        return Promise.async { loadFromThumbHash(bytes) }
    }
}

File: packages/react-native-nitro-image/ios/HybridImageFactory.swift
//
//  HybridImageFactory.swift
//  react-native-nitro-image
//
//  Created by Marc Rousavy on 10.06.25.
//

import Foundation
import NitroModules
import UniformTypeIdentifiers

class HybridImageFactory: HybridImageFactorySpec {
  /**
   * Load Image from file path
   */
  func loadFromFile(filePath rawFilePath: String) throws -> any HybridImageSpec {
    // 1. Clean out the file:// prefix
    let filePath = rawFilePath.replacingOccurrences(of: "file://", with: "")
    // 2. Load UIImage from file
    guard let uiImage = UIImage(contentsOfFile: filePath) else {
      throw RuntimeError.error(withMessage: "Failed to read image from file \"\(filePath)\"!")
    }
    return HybridImage(uiImage: uiImage)
  }
  func loadFromFileAsync(filePath: String) throws -> Promise<any HybridImageSpec> {
    return Promise.async {
      return try self.loadFromFile(filePath: filePath)
    }
  }

  /**
   * Load Image from resources
   */
  func loadFromResources(name: String) throws -> any HybridImageSpec {
    guard let uiImage = UIImage(named: name) else {
      throw RuntimeError.error(withMessage: "Image \"\(name)\" cannot be found in main resource bundle!")
    }
    return HybridImage(uiImage: uiImage)
  }
  func loadFromResourcesAsync(name: String) throws -> Promise<any HybridImageSpec> {
    return Promise.async {
      return try self.loadFromResources(name: name)
    }
  }

  /**
   * Load Image from SF Symbol Name
   */
  func loadFromSymbol(symbolName: String) throws -> any HybridImageSpec {
    guard let uiImage = UIImage(systemName: symbolName) else {
      throw RuntimeError.error(withMessage: "No Image with the symbol name \"\(symbolName)\" found!")
    }
    return HybridImage(uiImage: uiImage)
  }
  
  /**
   * Load Image from the given raw ArrayBuffer data
   */
  func loadFromRawPixelData(data: RawPixelData, allowGpu _ : Bool?) throws -> any HybridImageSpec {
    let uiImage = try UIImage(fromRawPixelData: data)
    return HybridImage(uiImage: uiImage)
  }
  
  func loadFromRawPixelDataAsync(data: RawPixelData, allowGpu: Bool?) throws -> Promise<any HybridImageSpec> {
    let dataCopy = data.buffer.isOwner ? data.buffer : ArrayBuffer.copy(of: data.buffer)
    let newData = RawPixelData(buffer: dataCopy,
                               width: data.width,
                               height: data.height,
                               pixelFormat: data.pixelFormat)
    return Promise.async {
      return try self.loadFromRawPixelData(data: newData, allowGpu: allowGpu)
    }
  }
  
  /**
   * Load Image from the given encoded ArrayBuffer data
   */
  func loadFromEncodedImageData(data: EncodedImageData) throws -> any HybridImageSpec {
    let copiedData = data.buffer.toData(copyIfNeeded: false)
    guard let uiImage = UIImage(data: copiedData) else {
      throw RuntimeError.error(withMessage: "The given ArrayBuffer could not be converted to a UIImage!")
    }
    return HybridImage(uiImage: uiImage)
  }
  
  func loadFromEncodedImageDataAsync(data: EncodedImageData) throws -> Promise<any HybridImageSpec> {
    let copiedData = data.buffer.toData(copyIfNeeded: true)
    return Promise.async {
      guard let uiImage = UIImage(data: copiedData) else {
        throw RuntimeError.error(withMessage: "The given ArrayBuffer could not be converted to a UIImage!")
      }
      return HybridImage(uiImage: uiImage)
    }
  }
  

  func loadFromThumbHash(thumbhash: ArrayBuffer) throws -> any HybridImageSpec {
    let data = thumbhash.toData(copyIfNeeded: false)
    let uiImage = thumbHashToImage(hash: data)
    return HybridImage(uiImage: uiImage)
  }
  func loadFromThumbHashAsync(thumbhash: ArrayBuffer) throws -> Promise<any HybridImageSpec> {
    let data = thumbhash.toData(copyIfNeeded: true)
    return Promise.async {
      let uiImage = thumbHashToImage(hash: data)
      return HybridImage(uiImage: uiImage)
    }
  }
}
